File: cpp/objects.cpp
// SPDX-FileCopyrightText: Copyright (c) 2022-2025 Yegor Bugayenko
// SPDX-License-Identifier: MIT

#include "./main.h"

class Bool {
public:
  virtual ~Bool() = default;
  virtual bool get() = 0;
};

class Int {
public:
  virtual ~Int() = default;
  virtual int get() = 0;
};

class Integer : public Int {
public:
  explicit Integer(int v) : value(v) {}
  explicit Integer(Int* i) : value(i->get()) {}
  ~Integer() override = default;
  int get() override { return value; }

private:
  int value;
};

class Less : public Bool {
public:
  Less(Int* l, Int* r) : left(l), right(r) {}
  ~Less() override {
    delete left;
    delete right;
  }
  bool get() override { return left->get() < right->get(); }

private:
  Int* left;
  Int* right;
};

class Add : public Int {
public:
  Add(Int* l, Int* r) : left(l), right(r) {}
  ~Add() override {
    delete left;
    delete right;
  }
  int get() override { return left->get() + right->get(); }

private:
  Int* left;
  Int* right;
};

class Sub : public Int {
public:
  Sub(Int* l, Int* r) : left(l), right(r) {}
  ~Sub() override {
    delete left;
    delete right;
  }
  int get() override { return left->get() - right->get(); }

private:
  Int* left;
  Int* right;
};

class If : public Int {
public:
  If(Bool* t, Int* l, Int* r) : term(t), left(l), right(r) {}
  ~If() override {
    delete term;
    delete left;
    delete right;
  }
  int get() override {
    if (term->get()) {
      return left->get();
    }
    return right->get();
  }

private:
  Bool* term;
  Int* left;
  Int* right;
};

class Fibo : public Int {
public:
  explicit Fibo(Int* v) : value(v) {}
  ~Fibo() override { delete value; }
  int get() override {
    Int* iff =
      new If(new Less(new Integer(value), new Integer(2)), new Integer(1),
        new Add(new Fibo(new Sub(new Integer(value), new Integer(1))),
          new Fibo(new Sub(new Integer(value), new Integer(2)))));
    const int result = iff->get();
    delete iff;
    return result;
  }

private:
  Int* value;
};

int calc(int x) {
  Int* fibo = new Fibo(new Integer(x));
  const int f = fibo->get();
  delete fibo;
  return f;
}

File: java/Objects.java
// SPDX-FileCopyrightText: Copyright (c) 2022-2025 Yegor Bugayenko
// SPDX-License-Identifier: MIT

class Objects {
  interface Bool {
    boolean get();
  }
  interface Int {
    int get();
  }
  static class Integer implements Int {
    private final int value;
    Integer(int v) {
      this.value = v;
    }
    @Override
    public int get() {
      return this.value;
    }
  }
  static class Less implements Bool {
    private final Int left;
    private final Int right;
    Less(Int l, Int r) {
      this.left = l;
      this.right = r;
    }
    @Override
    public boolean get() {
      return this.left.get() < this.right.get();
    }
  }
  static class Sub implements Int {
    private final Int left;
    private final Int right;
    Sub(Int l, Int r) {
      this.left = l;
      this.right = r;
    }
    @Override
    public int get() {
      return this.left.get() - this.right.get();
    }
  }
  static class Add implements Int {
    private final Int left;
    private final Int right;
    Add(Int l, Int r) {
      this.left = l;
      this.right = r;
    }
    @Override
    public int get() {
      return this.left.get() + this.right.get();
    }
  }
  static class If implements Int {
    private final Bool term;
    private final Int left;
    private final Int right;
    If(Bool t, Int l, Int r) {
      this.term = t;
      this.left = l;
      this.right = r;
    }
    @Override
    public int get() {
      if (this.term.get()) {
        return this.left.get();
      }
      return this.right.get();
    }
  }
  static class Fibo implements Int {
    private final Int value;
    Fibo(Int v) {
      this.value = v;
    }
    @Override
    public int get() {
      return new If(
        new Less(this.value, new Integer(2)),
        new Integer(1),
        new Add(
          new Fibo(new Sub(this.value, new Integer(1))),
          new Fibo(new Sub(this.value, new Integer(2)))
        )
      ).get();
    }
  }
  public static void main(final String... args) {
    if (args.length != 2) {
      System.out.print("Two args required: INPUT and CYCLES\n");
      return;
    }
    int total = 0;
    int f = 0;
    Integer input = new Integer(java.lang.Integer.parseInt(args[0]));
    int cycles = java.lang.Integer.parseInt(args[1]);
    for (int i = 0; i < cycles; ++i) {
      f = new Fibo(input).get();
      total += f;
    }
    System.out.print(
      String.format(
        "%d-th Fibonacci number is %d\nTotal is %d\n",
        input.get(), f, total
      )
    );
  }
}
